### ISSUE TITLE
dbc + tracepoint evaluated commands only show output of last evaluation

### ISSUE THREAD

**Issue by drvan** [radare/radare2#3377]
_TLDR: Inclusion of backtick evaluated expressions using dbc and tracepoints (note this issue doesn't occur with traditional breakpoints) causes the output to disappear for all except the last tracepoint hit. My uneducated guess is the output buffer isn't being properly flushed._

I'm attempting to view the evaluated value of what's in register EDX over the course of a loop. I've set a tracepoint at the location in question and set dbc as follows:

```
[0xf776da90]> db 0x080484c8
[0xf776da90]> dbte 0x080484c8
[0xf776da90]> "dbc 0x080484c8 ? `dr?ebx`"
[0xf776da90]> dc
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
Correct! Submit the password with your lab!
Debugging pid = 2059, tid = 1 now
84 0x54 0124 84 0000:0054 84 "T" 01010100 84.0 0.000000f 0.000000
```

Note that the evaluation of the register only shows on the last iteration of the loop, rather than at each iteration. To test this further, I changed dbc to only show the value of edx (no evaluation) and it worked as expected:

```
[0xf776cc10]> dbc 0x080484c8 dr?ebx
[0xf776cc10]> do
Wait event received by different pid 2059
Process with PID 2063 started...
File dbg://./4-1 YSAESAWTAHT reopened in read-write mode
Debugging pid = 2063, tid = 2063 now
Assuming filepath ./4-1
Attached debugger to pid = 2063, tid = 2063
[0xf776fa90]> dc
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
Correct! Submit the password with your lab!

==> Process finished

Child 2063 is dead
0x00000059
0x00000053
0x00000041
0x00000045
0x00000053
0x00000041
0x00000057
0x00000054
0x00000041
0x00000048
0x00000054
0x00000054
```

This is what I expected (except not sure why the output is showing up AFTER the tracepoints have been hit, and not after each tracepoint, but that's beside the point). 

To rule out the evaluation operator being the culprit, I changed dbc to a straight constant evaluation and repeated the run:

```
[0x080484c8]> dbc 0x080484c8 ? 0x00000059
[0x080484c8]> do
Process with PID 2064 started...
File dbg://./4-1 YSAESAWTAHT reopened in read-write mode
Debugging pid = 2064, tid = 2064 now
Assuming filepath ./4-1
Attached debugger to pid = 2064, tid = 2064
[0xf76eaa90]> dc
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
Correct! Submit the password with your lab!

==> Process finished

Child 2064 is dead
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
```

Here, the evaluation worked successfully. As a final test, I used the same command, this time with the entire thing quoted, to see if quotes were the problem:

```
[0x080484c8]> "dbc 0x080484c8 ? 0x00000059"
[0x080484c8]> do
Process with PID 2066 started...
File dbg://./4-1 YSAESAWTAHT reopened in read-write mode
Debugging pid = 2066, tid = 2066 now
Assuming filepath ./4-1
Attached debugger to pid = 2066, tid = 2066
[0xf7738a90]> dc
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
hit tracepoit at: 80484c8
Correct! Submit the password with your lab!

==> Process finished

Child 2066 is dead
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
89 0x59 0131 89 0000:0059 89 "Y" 01011001 89.0 0.000000f 0.000000
```

As seen, quotes don't appear to be the issue. 

Platform info:

```
$ r2 -v
WARNING: r2 library versions mismatch! See r2 -V
radare2 0.10.0-git 9109 @ linux-little-x86-64 git.0.9.9-1004-gda32642
commit: da3264250c628f093a790758f8985823af25ae66 build: 2015-09-26
$ uname -a
Linux vagrant 3.19.0-26-generic #28-Ubuntu SMP Tue Aug 11 14:16:32 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -a
LSB Version:    core-2.0-amd64:core-2.0-noarch:core-3.0-amd64:core-3.0-noarch:core-3.1-amd64:core-3.1-noarch:core-3.2-amd64:core-3.2-noarch:core-4.0-amd64:core-4.0-noarch:core-4.1-amd64:core-4.1-noarch:security-4.0-amd64:security-4.0-noarch:security-4.1-amd64:security-4.1-noarch
Distributor ID: Ubuntu
Description:    Ubuntu 15.04
Release:    15.04
Codename:   vivid
```

I can provide the binary, binary source code, and compilation options if needed.

**Comment by radare** [comment]
Yeah rconsflush should be called right after every hit. Tracing is not suposed to be buffered in a shot. Otherwise we should just dump to stderr.

Thanks for pointing out this. Will fix soon

**Comment by puxlit** [comment]
Wrapping the commands in a macro seems to reduce (though not quite eliminate) the delay.

```
[0x7fff98834f46]> db sym._setCallback
[0x7fff98834f46]> dbte sym._setCallback
[0x7fff98834f46]> "dbc sym._setCallback ? `dr?rax`"
[0x7fff98834f46]> dc
hit tracepoint at: 10e601a60
hit tracepoint at: 10e601a60
hit tracepoint at: 10e601a60
hit tracepoint at: 10e601a60
^Cmessage interrupted
19 0x13 023 19 0000:0013 19 "\x13" 00010011 19.0 19.000000f 19.000000
20 0x14 024 20 0000:0014 20 "\x14" 00010100 20.0 20.000000f 20.000000
21 0x15 025 21 0000:0015 21 "\x15" 00010101 21.0 21.000000f 21.000000
22 0x16 026 22 0000:0016 22 "\x16" 00010110 22.0 22.000000f 22.000000
[0x7fff98834f46]> (lol,? `dr?rax`)
[0x7fff98834f46]> dbc sym._setCallback .(lol)
[0x7fff98834f46]> dc
hit tracepoint at: 10e601a60
hit tracepoint at: 10e601a60
23 0x17 027 23 0000:0017 23 "\x17" 00010111 23.0 23.000000f 23.000000
hit tracepoint at: 10e601a60
24 0x18 030 24 0000:0018 24 "\x18" 00011000 24.0 24.000000f 24.000000
hit tracepoint at: 10e601a60
25 0x19 031 25 0000:0019 25 "\x19" 00011001 25.0 25.000000f 25.000000
hit tracepoint at: 10e601a60
26 0x1a 032 26 0000:001a 26 "\x1a" 00011010 26.0 26.000000f 26.000000
^Cmessage interrupted
27 0x1b 033 27 0000:001b 27 "\x1b" 00011011 27.0 27.000000f 27.000000
```

**Comment by puxlit** [comment]
This is good to close; #7752 did the trick.
